PR Script

rambleraptor opened the pull request:
As part of my dev machine debacle, I accidentally deleted my old PR script. That might've actually been for the best.

This script creates PRs on Ansible Core for all of the different modules. You can only create 1 PR per module, so if we make a change across all modules, we have to send `all modules` # of PRs. If there's changes on certain helper files, those will be sent along as well.

Currently, I run this script locally. Eventually, I'd like to build a Concourse job that runs this script and does weekly 'releases'


<!--
Changes per downstream repository.  For each repository that you
expect to have changed, find the [tag] and write your commit
message beneath it.  More-specific tags replace less-specific tags.
For example, if you provide a message under [all], a message under
[puppet], and a message under [puppet-dns], the Terraform repository
will have the resulting commit made using the [all] message, the
Puppet Compute repository will have its commit made using [puppet],
and the Puppet DNS repository will have its commit made using
[puppet-dns].  You can delete unused tags, but you don't need to.

The structure of the PR body is important to our CI system!
The comments can be deleted, but if you want to make the downstream
commits sensible, you'll need to leave the dashed line separating
this PR's changes from the commit messages for downstream commits.
-->

-----------------------------------------------------------------
# [all]
## [terraform]
## [puppet]
### [puppet-bigquery]
### [puppet-compute]
### [puppet-container]
### [puppet-dns]
### [puppet-logging]
### [puppet-pubsub]
### [puppet-resourcemanager]
### [puppet-sql]
### [puppet-storage]
## [chef]
### [chef-compute]
### [chef-container]
### [chef-dns]
### [chef-logging]
### [chef-spanner]
### [chef-sql]
### [chef-storage]
## [ansible]

modular-magician commented:
I am a robot that works on MagicModules PRs!

I built this PR into one or more PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/GoogleCloudPlatform/chef-google-compute/pull/48

danawillow commented:
I'm removing myself as a reviewer because I'm not super good at git and probably wouldn't have much to say, but if you put me here because you do want my opinion explicitly, feel free to add me back!

rambleraptor commented:
Sounds good! I didn't really know who should be placed on this, so I put everyone.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 9cceebc0) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 95ad99fe) have been included in your existing downstream PRs.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 3b083d05) have been included in your existing downstream PRs.

rambleraptor commented:
@ndmckinley @SirGitsalot ptal.

I changed up the script pretty substantially to use the new Ansible system (one PR for all bug fixes, one PR for each new module) and to give the bug fix PRs a changelog (using the labels I added in on the Magician)

ndmckinley commented on tools/ansible-pr/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## Requirements
+* `hub` CLI
+* `build/ansible` has the origin remote to whatever fork of ansible
```
Still think this phrasing is a little confusing.

rambleraptor replied:
Changed.

ndmckinley commented on tools/ansible-pr/run:
```diff
@@ -0,0 +1,230 @@
+  if @options[:dry_run]
+    puts command.join('; ')
+  else
+    %x(#{command.join('; ')})
```
Is there no better way than shelling out here?  I strongly prefer a ruby library that manages git repos for you.  Without that, I wonder if it'd be a better shell script than ruby script.

rambleraptor replied:
I spent way too much time trying to work with the [rugged gem](https://github.com/libgit2/rugged).

Issues I had included:
* Getting it to work with SSH repos was incredibly hard.
* Documentation isn't great.
* It can replicate most of this script, but not all of it (finding all files not in x branch).

I'm not convinced the gem can get the job done or that it's worth the time trying to figure out if it can get the job done.

As for converting to a Bash script, how strong of an opinion do you have? My bias is that simple scripts should be Bash, complicated should be anything else. I think this script is complicated enough to warrant being in anything-but-Bash.

SirGitsalot replied:
If we could use gbash that would make things not so bad, but since we're already in ruby here I think it makes sense.

My concern with shelling out from a ruby script is the same as always: will this ever be run any place other than the user's own machine, e.g., on concourse? If not, I'm not too worried about an attacker naming their branch `foo; rm -rf ~ #` since they'll just be owning themselves.

ndmckinley commented on tools/ansible-pr/run:
```diff
@@ -0,0 +1,230 @@
+    @options[:dry_run] = true
+  end
+
+  opts.on('--date [DATE (as ISO8601 String)]', 'Date for Github sorting') do |v|
```
Could you elaborate on this in the help message?  I don't really understand what it means.

rambleraptor replied:
Done.

rambleraptor commented:
@SirGitsalot @ndmckinley ptal. This is a final version that works with all of Ansible's rules. Now that the review/merge cycle is shortened, I can finally have confidence that this script actually complies with Ansible's requirements + ours.

ndmckinley commented:
Well, hm.  I mean, I'm not a big fan of the `system`-heavy situation here, but I'm not really going to be using it, so I'm not sure I should really be holding this up based on my preferences.  I'm cool with it going in if @SirGitsalot is.

rambleraptor commented:
I've had experiences where the Bash script grew so unwieldy that it was impossible for anybody other than the creator to understand it. That's my biggest worry. I'm not a huge fan of all the interspersed system() calls either.

rambleraptor commented:
@ndmckinley @SirGitsalot can I get some consensus on what we would like to do with this? I'd like to build out a pipeline using this script to get us nightly releases.

ndmckinley commented:
Same as before - it's still not how I'd do it personally, but if you won the lottery and it became someone else's job to keep this thing running, they would definitely be able to do it.  I say merge and get building.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit d9f4c970) have been included in your existing downstream PRs.
